File: FSUIKit/Classes/share/FSShare.m
//
//  FSShare.m
//  myhome
//
//  Created by FudonFuchina on 2017/8/25.
//  Copyright © 2017年 fuhope. All rights reserved.
//

#import "FSShare.h"
#import <MessageUI/MessageUI.h>

#import "FSKit.h"
#import "FSUIKit.h"
#import "FSImage.h"

#import "FSUIKit-Swift.h"

@interface FSShare ()<MFMessageComposeViewControllerDelegate,MFMailComposeViewControllerDelegate,UIDocumentInteractionControllerDelegate>

@property (nonatomic, strong) UIDocumentInteractionController   *documentController;

@end

@implementation FSShare

static FSShare *_instance = nil;
+(FSShare *)sharedInstance{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[FSShare alloc] init];
    });
    return _instance;
}

+ (void)wechatAPIRegisterAppKey:(NSString *)appKey{
#if TARGET_IPHONE_SIMULATOR
#else
    if ([appKey isKindOfClass:NSString.class] && appKey.length) {
        [WXApi registerApp:appKey];
    }
#endif
}

+ (BOOL)handleOpenUrl:(NSURL *)url{
    FSShare *share = [[FSShare alloc] init];
    return [WXApi handleOpenURL:url delegate:share];
}

#pragma mark WXDelegate
-(void)onReq:(BaseReq*)req {
}

-(void)onResp:(BaseResp*)resp {
    if([resp isKindOfClass:SendMessageToWXResp.class]){   // SendAuthResp
        UIWindowScene *ws = [FSKit currentWindowScene];
        if (@available(iOS 15.0, *)) {
            [FSUIKit showAlertWithMessage: resp.errStr controller: ws.keyWindow.rootViewController];
        }
    }
}

// 微信分享图片
+ (void)wxImageShareActionWithImage:(UIImage *)image controller:(UIViewController *)controller result:(void(^)(NSString *bResult))completion{
    if (!([image isKindOfClass:[UIImage class]] && image.size.width > 1 && image.size.height > 1)) {
        [FSUIKit showAlertWithMessage: @"分享的不是图片" controller: controller];
        return;
    }
    BOOL hasInstalled = [self checkPhoneHasWechat: controller];
    if (!hasInstalled) {
        return;
    }
        
    UIImage *thumbImage = [FSImage compressImage:image width:50];
    
    WXMediaMessage *message = [WXMediaMessage message];
    [message setThumbImage:thumbImage];
    
    WXImageObject *ext = [WXImageObject object];
    ext.imageData = UIImagePNGRepresentation(image);
    message.mediaObject = ext;
    
    SendMessageToWXReq *req = [[SendMessageToWXReq alloc] init];
    req.bText = NO;
    req.message = message;
    req.scene = WXSceneSession;
    [WXApi sendReq:req];
}

// 微信文件分享
+ (void)wxFileShareActionWithPath:(NSString *)path fileName:(NSString *)fileName extension:(NSString *)extension controller:(UIViewController *)controller result:(void(^)(NSString *bResult))completion{
    NSFileManager *manager = [NSFileManager defaultManager];
    if ((![manager fileExistsAtPath:path]) || fileName == nil || extension == nil) {
        return;
    }
    BOOL hasInstalled = [self checkPhoneHasWechat: controller];
    if (!hasInstalled) {
        return;
    }
    WXFileObject *ext = [WXFileObject object];
    ext.fileExtension = extension;
    ext.fileData = [NSData dataWithContentsOfFile:path];
    
    WXMediaMessage *message = [WXMediaMessage message];
    message.title = [[NSString alloc] initWithFormat:@"%@.%@",fileName,extension];
    message.description = @"文件分享";
    //    [message setThumbImage:[UIImage imageNamed:@"res2.jpg"]];
    message.mediaObject = ext;
    
    SendMessageToWXReq *req = [[SendMessageToWXReq alloc] init];
    req.bText = NO;
    req.message = message;
    req.scene = WXSceneSession;
    [WXApi sendReq:req];
}

+ (void)wxContentShare:(NSString *)content scene:(int)scene controller:(UIViewController *)controller {
    if (!_fs_isValidateString(content)) {
        return;
    }
    BOOL hasInstalled = [self checkPhoneHasWechat: controller];
    if (!hasInstalled) {
        return;
    }
    SendMessageToWXReq *req = [[SendMessageToWXReq alloc] init];
    req.text = content;
    req.bText = YES;
    req.scene = scene;
    [WXApi sendReq:req];
}

+ (void)wxUrlShareTitle:(NSString *)title description:(NSString *)description url:(NSString *)url controller:(UIViewController *)controller {
    BOOL hasInstalled = [self checkPhoneHasWechat: controller];
    if (!hasInstalled) {
        return;
    }
    WXWebpageObject *ext = [WXWebpageObject object];
    ext.webpageUrl = url;
 
    WXMediaMessage *message = [WXMediaMessage message];
    message.title = title;
    message.description = description;
    [message setThumbImage:[UIImage imageNamed:@"logo_wx_small"]];
    message.mediaObject = ext;
    
    SendMessageToWXReq* req = [[SendMessageToWXReq alloc] init];
    req.bText = NO;
    req.message = message;
    req.scene = WXSceneSession;
    [WXApi sendReq:req];
}

+ (BOOL)checkPhoneHasWechat:(UIViewController *)controller {
    BOOL hasInstalledWeChat = [WXApi isWXAppInstalled];
    if (!hasInstalledWeChat) {
        [FSUIKit alert: UIAlertControllerStyleAlert controller: controller title: @"微信分享" message: @"未安装微信，是否去下载？" actionTitles: @[@"下载"] styles: @[@(UIAlertActionStyleDefault)] handler:^(FSAlertAction *action) {
            NSString *url = [WXApi getWXAppInstallUrl];
            NSURL *u = [NSURL URLWithString: url];
            if ([UIApplication.sharedApplication canOpenURL: u]) {
                if (@available(iOS 10.0, *)) {
                    [UIApplication.sharedApplication openURL: u options: @{} completionHandler:^(BOOL success) {}];
                }
            }
        }];
    }
    return hasInstalledWeChat;
}

// 邮件分享
+ (void)emailShareWithSubject:(NSString *)subject on:(UIViewController *)c messageBody:(NSString *)body recipients:(NSArray *)recipients{
    [self emailShareWithSubject:subject on:c messageBody:body recipients:recipients fileData:nil fileName:nil mimeType:nil];
}

+ (void)emailShareWithSubject:(NSString *)subject on:(UIViewController *)c messageBody:(NSString *)body recipients:(NSArray *)recipients fileData:(NSData *)data fileName:(NSString *)fileName mimeType:(NSString *)fileType{
    [[FSShare sharedInstance] emailShareWithSubject:subject on:c messageBody:body recipients:recipients fileData:data fileName:fileName mimeType:fileType ];
}

- (void)emailShareWithSubject:(NSString *)subject on:(UIViewController *)controller  messageBody:(NSString *)body recipients:(NSArray *)recipients fileData:(NSData *)data fileName:(NSString *)fileName mimeType:(NSString *)fileType {
    if (![controller isKindOfClass:UIViewController.class]) {
        return;
    }
    MFMailComposeViewController *picker = [[MFMailComposeViewController alloc] init];
    picker.mailComposeDelegate = self;
    if (_fs_isValidateString(subject)) {
        [picker setSubject:subject];
    }
    if (_fs_isValidateArray(recipients)) {
        [picker setToRecipients:recipients];
    }
    if (body) {
        [picker setMessageBody:body isHTML:NO];
    }
    if ([data isKindOfClass:NSData.class] && data.length) {
        [picker addAttachmentData:data mimeType:fileType fileName:fileName];
    }
    if (picker) {
        [controller presentViewController:picker animated:YES completion:nil];
    }
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error{
    NSString *msg = nil;
    switch (result){
        case MFMailComposeResultCancelled:
//            msg = @"邮件发送取消";
            break;
        case MFMailComposeResultSaved:
            msg = @"邮件保存成功";
            break;
        case MFMailComposeResultSent:
            msg = @"邮件推送成功";
            break;
        case MFMailComposeResultFailed:
            msg = @"邮件发送失败";
            break;
        default:
            break;
    }
    
    if (msg) {
        __unused id obj = [FSToastS toast: msg duration: 2];
    }
    [controller dismissViewControllerAnimated: YES completion: nil];
}

// 短信分享
+ (void)messageShareWithMessage:(NSString *)message on:(UIViewController *)controller recipients:(NSArray *)recipients{
    [self messageShareWithMessage:message on:controller recipients:recipients data:nil fileName:nil fileType:nil];
}

+ (void)messageShareWithMessage:(NSString *)message on:(UIViewController *)controller recipients:(NSArray *)recipients data:(NSData *)fileData fileName:(NSString *)fileName fileType:(NSString *)fileType{    // @"image/jpeg"    @"text/txt"     @"text/doc"     @"file/pdf"
    [[FSShare sharedInstance] messageShareWithMessage:message on:controller recipients:recipients data:fileData fileName:fileName fileType:fileType];
}

- (void)messageShareWithMessage:(NSString *)message on:(UIViewController *)controller recipients:(NSArray *)recipients data:(NSData *)fileData fileName:(NSString *)fileName fileType:(NSString *)fileType{    // @"image/jpeg"    @"text/txt"     @"text/doc"     @"file/pdf"
    if (![controller isKindOfClass:UIViewController.class]) {
        return;
    }
    if (![MFMessageComposeViewController canSendText]) {
        [FSUIKit showAlertWithMessage:@"设备不支持发送短信" controller:controller];
        return;
    }
    MFMessageComposeViewController *picker = [[MFMessageComposeViewController alloc] init];
    picker.messageComposeDelegate = self;
    if (recipients) {
        picker.recipients = recipients;
    }
    if (message) {
        picker.body = message;
    }
    if ([fileData isKindOfClass:NSData.class]) {
        if (!([fileName isKindOfClass:NSString.class] && fileName.length)) {
            fileName = @((NSInteger)[[NSDate date] timeIntervalSince1970]).stringValue;
        }
        [picker addAttachmentData:fileData typeIdentifier:fileType filename:fileName];
    }
    if (picker) {
        [controller presentViewController:picker animated:YES completion:nil];
    }
}

- (void)messageComposeViewController:(MFMessageComposeViewController *)controller didFinishWithResult:(MessageComposeResult)result{
    NSString *msg = nil;
    switch (result){
        case MessageComposeResultCancelled:
//            msg = @"短信发送取消";
            break;
        case MessageComposeResultSent:
            msg = @"短信发送成功";
            break;
        case MessageComposeResultFailed:
            msg = @"短信发送失败";
            break;
        default:
            break;
    }
    
    if (msg) {
        __unused id obj = [FSToastS toast: msg duration: 2];
    }
    [controller dismissViewControllerAnimated: YES completion: nil];
}

- (void)openUIDocumentInteractionController:(NSURL *)fileURL inController:(UIViewController *)controller {
    if (fileURL) {
        if (!self.documentController) {
            self.documentController = [UIDocumentInteractionController interactionControllerWithURL:fileURL];
            [self.documentController setDelegate:self];
        }else{
            self.documentController.URL = fileURL;
        }
        BOOL canOpen =  [self.documentController presentOpenInMenuFromRect:CGRectZero inView:controller.view animated:YES];
        if (canOpen == NO) {
            [FSUIKit showAlertWithMessage:@"出现问题，不能打开" controller:controller];
        }
    }
}

- (void)documentInteractionControllerDidDismissOpenInMenu:(UIDocumentInteractionController *)controller {
    self.documentController = nil;
}


@end

File: FSUIKit/Classes/Swift/FSShare.swift
// FSShare.swift
// Translated from FSShare.h/m
// Note: 微信SDK相关功能需要集成WechatOpenSDK

import UIKit
import MessageUI

//import WechatOpenSDK  暂不支持，需要OC桥接，后续实现

@objcMembers
public class FSShare: NSObject {
    
    public static let shared = FSShare()
    
    public var result: ((Bool, String?) -> Void)?
    
    private var documentController: UIDocumentInteractionController?
    
    private override init() {
        super.init()
    }
    
    // MARK: - 微信注册
    
    public static func wechatAPIRegisterAppKey(_ appKey: String?) {
//        #if !targetEnvironment(simulator)
//        guard let appKey = appKey, !appKey.isEmpty else { return }
//        WXApi.registerApp(appKey, universalLink: "")
//        #endif
    }
    
    public static func handleOpenUrl(_ url: URL) -> Bool {
        return false
//        let share = FSShare()
//        return WXApi.handleOpen(url, delegate: share)
    }
    
    // MARK: - 微信分享图片
    
    public static func wxImageShareAction(with image: UIImage?, controller: UIViewController, result completion: ((String?) -> Void)?) {
//        guard let image = image, image.size.width > 1, image.size.height > 1 else {
//            FSUIKit.showAlert(withMessage: "分享的不是图片", controller: controller)
//            return
//        }
//        
//        guard checkPhoneHasWechat(controller) else { return }
//        
//        let thumbImage = FSImage.compressImage(image, width: 50)
//        
//        let message = WXMediaMessage()
//        message.setThumbImage(thumbImage)
//        
//        let ext = WXImageObject()
//        ext.imageData = image.pngData()
//        message.mediaObject = ext
//        
//        let req = SendMessageToWXReq()
//        req.bText = false
//        req.message = message
//        req.scene = Int32(WXSceneSession.rawValue)
//        WXApi.send(req)
    }
    
    // MARK: - 微信文件分享
    
    public static func wxFileShareAction(withPath path: String?, fileName: String?, extension ext: String?, controller: UIViewController, result completion: ((String?) -> Void)?) {
//        guard let path = path, let fileName = fileName, let ext = ext else { return }
//        
//        let manager = FileManager.default
//        guard manager.fileExists(atPath: path) else { return }
//        guard checkPhoneHasWechat(controller) else { return }
//        
//        let fileObject = WXFileObject()
//        fileObject.fileExtension = ext
//        fileObject.fileData = try? Data(contentsOf: URL(fileURLWithPath: path))
//        
//        let message = WXMediaMessage()
//        message.title = "\(fileName).\(ext)"
//        message.description = "文件分享"
//        message.mediaObject = fileObject
//        
//        let req = SendMessageToWXReq()
//        req.bText = false
//        req.message = message
//        req.scene = Int32(WXSceneSession.rawValue)
//        WXApi.send(req)
    }
    
    // MARK: - 微信文字分享
    
    public static func wxContentShare(_ content: String?, scene: Int32, controller: UIViewController) {
        guard let content = content, !content.isEmpty else { return }
        guard checkPhoneHasWechat(controller) else { return }
        
//        let req = SendMessageToWXReq()
//        req.text = content
//        req.bText = true
//        req.scene = scene
//        WXApi.send(req)
    }
    
    // MARK: - 微信URL分享
    
    public static func wxUrlShareTitle(_ title: String?, description: String?, url: String?, controller: UIViewController) {
        guard checkPhoneHasWechat(controller) else { return }
        
//        let ext = WXWebpageObject()
//        ext.webpageUrl = url ?? ""
//        
//        let message = WXMediaMessage()
//        message.title = title
//        message.description = description
//        message.setThumbImage(UIImage(named: "logo_wx_small"))
//        message.mediaObject = ext
//        
//        let req = SendMessageToWXReq()
//        req.bText = false
//        req.message = message
//        req.scene = Int32(WXSceneSession.rawValue)
//        WXApi.send(req)
    }
    
    // MARK: - 检查微信安装
    
    public static func checkPhoneHasWechat(_ controller: UIViewController) -> Bool {
//        let hasInstalled = WXApi.isWXAppInstalled()
//        if !hasInstalled {
//            FSUIKit.alert(.actionSheet, controller: controller, title: "微信分享", message: "未安装微信，是否去下载？", actionTitles: ["下载"], styles: [NSNumber(value: UIAlertAction.Style.default.rawValue)]) { action in
//                if let url = URL(string: WXApi.getWXAppInstallUrl()) {
//                    if UIApplication.shared.canOpenURL(url) {
//                        UIApplication.shared.open(url, options: [:], completionHandler: nil)
//                    }
//                }
//            }
//        }
//        return hasInstalled
        
        return false
    }
    
    // MARK: - 短信分享
    
    public static func messageShare(withMessage message: String?, on controller: UIViewController, recipients: [String]?) {
        messageShare(withMessage: message, on: controller, recipients: recipients, data: nil, fileName: nil, fileType: nil)
    }
    
    public static func messageShare(withMessage message: String?, on controller: UIViewController, recipients: [String]?, data fileData: Data?, fileName: String?, fileType: String?) {
        shared.messageShare(withMessage: message, on: controller, recipients: recipients, data: fileData, fileName: fileName, fileType: fileType)
    }
    
    public func messageShare(withMessage message: String?, on controller: UIViewController, recipients: [String]?, data fileData: Data?, fileName: String?, fileType: String?) {
        guard MFMessageComposeViewController.canSendText() else {
            FSUIKit.showAlertWithMessage("设备不支持发送短信", controller: controller)
            return
        }
        
        let picker = MFMessageComposeViewController()
        picker.messageComposeDelegate = self
        
        if let recipients = recipients {
            picker.recipients = recipients
        }
        if let message = message {
            picker.body = message
        }
        if let fileData = fileData {
            let name = (fileName?.isEmpty == false) ? fileName! : "\(Int(Date().timeIntervalSince1970))"
            picker.addAttachmentData(fileData, typeIdentifier: fileType ?? "", filename: name)
        }
        
        controller.present(picker, animated: true, completion: nil)
    }
    
    // MARK: - 邮件分享
    
    public static func emailShare(withSubject subject: String?, on controller: UIViewController, messageBody body: String?, recipients: [String]?) {
        emailShare(withSubject: subject, on: controller, messageBody: body, recipients: recipients, fileData: nil, fileName: nil, mimeType: nil)
    }
    
    public static func emailShare(withSubject subject: String?, on controller: UIViewController, messageBody body: String?, recipients: [String]?, fileData data: Data?, fileName: String?, mimeType fileType: String?) {
        shared.emailShare(withSubject: subject, on: controller, messageBody: body, recipients: recipients, fileData: data, fileName: fileName, mimeType: fileType)
    }
    
    public func emailShare(withSubject subject: String?, on controller: UIViewController, messageBody body: String?, recipients: [String]?, fileData data: Data?, fileName: String?, mimeType fileType: String?) {
        guard MFMailComposeViewController.canSendMail() else {
            FSUIKit.showAlertWithMessage("设备不支持发送邮件", controller: controller)
            return
        }
        
        let picker = MFMailComposeViewController()
        picker.mailComposeDelegate = self
        
        if let subject = subject, !subject.isEmpty {
            picker.setSubject(subject)
        }
        if let recipients = recipients, !recipients.isEmpty {
            picker.setToRecipients(recipients)
        }
        if let body = body {
            picker.setMessageBody(body, isHTML: false)
        }
        if let data = data, !data.isEmpty, let fileName = fileName, let fileType = fileType {
            picker.addAttachmentData(data, mimeType: fileType, fileName: fileName)
        }
        
        controller.present(picker, animated: true, completion: nil)
    }
    
    // MARK: - UIDocumentInteractionController
    
    public func openDocumentInteractionController(_ fileURL: URL?, in controller: UIViewController) {
        guard let fileURL = fileURL else { return }
        
        if documentController == nil {
            documentController = UIDocumentInteractionController(url: fileURL)
            documentController?.delegate = self
        } else {
            documentController?.url = fileURL
        }
        
        let canOpen = documentController?.presentOpenInMenu(from: .zero, in: controller.view, animated: true) ?? false
        if !canOpen {
            FSUIKit.showAlertWithMessage("出现问题，不能打开", controller: controller)
        }
    }
}

// MARK: - WXApiDelegate

//extension FSShare: WXApiDelegate {
//    public func onReq(_ req: BaseReq) {
//        // Handle request
//    }
//    
//    public func onResp(_ resp: BaseResp) {
//        if resp is SendMessageToWXResp {
//            if let windowScene = FSKit.currentWindowScene(),
//               let rootVC = windowScene.keyWindow?.rootViewController {
//                FSUIKit.showAlert(withMessage: resp.errStr, controller: rootVC)
//            }
//        }
//    }
//}

// MARK: - MFMessageComposeViewControllerDelegate

extension FSShare: MFMessageComposeViewControllerDelegate {
    public func messageComposeViewController(_ controller: MFMessageComposeViewController, didFinishWith result: MessageComposeResult) {
        var msg: String?
        switch result {
        case .cancelled:
            break
        case .sent:
            msg = "短信发送成功"
        case .failed:
            msg = "短信发送失败"
        @unknown default:
            break
        }
        
        if let msg = msg {
            _ = FSToast.toast(msg, duration: 2)
        }
        controller.dismiss(animated: true, completion: nil)
    }
}

// MARK: - MFMailComposeViewControllerDelegate

extension FSShare: MFMailComposeViewControllerDelegate {
    public func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        var msg: String?
        switch result {
        case .cancelled:
            break
        case .saved:
            msg = "邮件保存成功"
        case .sent:
            msg = "邮件推送成功"
        case .failed:
            msg = "邮件发送失败"
        @unknown default:
            break
        }
        
        if let msg = msg {
            _ = FSToast.toast(msg, duration: 2)
        }
        controller.dismiss(animated: true, completion: nil)
    }
}

// MARK: - UIDocumentInteractionControllerDelegate

extension FSShare: UIDocumentInteractionControllerDelegate {
    public func documentInteractionControllerDidDismissOpenInMenu(_ controller: UIDocumentInteractionController) {
        documentController = nil
    }
}
